Start with data stream lifecycle documentation (#95326)
elastic/elasticsearch: docs/reference/dlm/index.asciidoc

[role="xpack"]
[[data-lifecycle-management]]
== {dlm-init}: Manage your data lifecycle

preview::[]

{dlm-cap} ({dlm-init}) enables you to automatically manage data streams
according to your retention requirements. For example, you could use {dlm-init} to:

* Ensure that data indexed in a data stream will be kept at least for the retention time you defined.

> * Older data will be deleted by {es} any moment later in time considering {es}'s performance.

Reviewer: ```suggestion
* Older data will be deleted by {es} at a later time.
```

Not sure what the es performance means here? Maybe we should skip it?

> You can create a data stream via index templates and configure your data lifecycle as part of the index template or you

Reviewer: Should we talk here about the data stream lifecycle _configuration_  for new and existing data streams as opposed to creating a data stream? (it's not technically correct that you create a data stream using the index templates but more so that a write request that matches an index pattern in some templates will trigger the creation of the data stream)

Reviewer: Before moving into the tutorials, should we describe some of the goals and internals of DLM? 
e.g. that it runs in a loop based on a setting, that it only applies to data streams,  it looks to roll over the write index of the data stream and to execute the retention for the other backing indices 
I guess something similar to the ILM overview https://www.elastic.co/guide/en/elasticsearch/reference/current/overview-index-lifecycle-management.html

What do you think?

Reviewer: I think we should also talk here a bit about how the `lifecycle` bit of an index template is only applied to NEW data streams (as opposed to the other parts of the `template` like `settings` and `mappings` that are applied to every new _backing index_ of a data stream) i.e. on a data stream rollover the `lifecycle` will NOT be configured on the data stream

I think this is quite a fundamental difference in the properties of a `template` and we should make the behaviour very clear - I'd say here and in the index template docs

What do you think?

Author: I agree with all of the above, I will work on it.

[role="xpack"]
[[data-lifecycle-management]]
== {dlm-init}: Manage your data lifecycle

preview::[]

The goal of {dlm} ({dlm-init}) is to enable you to automatically manage data streams

> according to your retention requirements. For example, you could use {dlm-init} to:

Reviewer 2: `automatically manage data streams according to your retention requirements, without having to worry about...` complicated low-level configuration? what happens which each individual index? I'm not really sure. Just seems like this sentence undersells DLM :)

Author: I see what you mean. I do not want to mention indices here, or other implementation details. If we have new users that do not know ILM then this comparison will not be relevant. I can stress a bit more that it's "easy" and leave it at that.

In follow up documentation that would explain better the differences between ILM and DLM this comparison would be more appropriate. What do you think?

Reviewer 3: I also paused on this sentence (in today's form where it ends with "without having to worry." I think this still needs some rewording as in this form I'm left thinking "worry about what?". I think what we mean is "without having to worry about internal implementation details." or "without having to consider as many details as with ILM."?

Author: I see, it was meant to be that it will just work :p . I will remove it. Let's leave comparisons for later.

* Ensure that data indexed in a data stream will be kept at least for the retention time you defined.

> * Older data will be deleted by {es} at a later time.

Reviewer 2: This second bullet doesn't seem to be parallel with the first, and doesn't make sense following `you could use DLM to`. Maybe something like `Ensure that data older than your retention time will be deleted automatically` or something?

Author: Thanks! You are absolutely right!

To achieve that, it supports:

* <<index-rollover,Rollover>>: which chunks your incoming data in smaller pieces to facilitate better performance
and backwards incompatible mapping changes.
* Configurable retention: allows you to configure the time period for which your data is guaranteed to be stored.
{es} is allowed at a later time to delete data older than this time period.

[discrete]
[[dlm-how-it-works]]
=== How does {dlm-init} work?

{dlm-init} runs in intervals configured by <<indices-dlm-poll-interval,`indices.dlm.poll_interval`>>. At every run,
{dlm-init} goes over each data stream and performs the following steps:

. Checks if the data stream is managed by {dlm-init}, meaning it has a data lifecycle configured. If the data stream is
managed by {dlm-init}, then

> . it will try to rollover the write backing index of the data stream, given that the backing index is

Author: I think both are good ideas. I will address them.

To achieve that, it supports:

* <<index-rollover,Rollover>>: which chunks your incoming data in smaller pieces to facilitate better performance
and backwards incompatible mapping changes.
* Configurable retention: allows you to configure the time period for which your data is guaranteed to be stored.
{es} is allowed at a later time to delete data older than this time period.

[discrete]
[[dlm-how-it-works]]
=== How does {dlm-init} work?

{dlm-init} runs in intervals configured by <<indices-dlm-poll-interval,`indices.dlm.poll_interval`>>. At every run,
{dlm-init} goes over each data stream and performs the following steps:

. Checks if the data stream is managed by {dlm-init}, meaning it has a data lifecycle configured. If the data stream is
managed by {dlm-init}, then
. it will rollover the write backing index of the data stream, given it fulfills the conditions defined by
<<cluster-dlm-default-rollover,`cluster.dlm.default.rollover`>>,
. it will apply retention to the remaining backing indices. This means that it will delete  the backing indices that
have completed the retention period after they have been rolled over. The time after rollover is called `generation_time`

> and it ensures that all the data in this index are at least older than this time.

Author: We are mentioning already in a previous part of this page and at another spot that the data gets deleted "at a later time". However, since we are explaining here how things work, I elaborated a bit more on the rationale behind the generation time and deletion. Let me know what you think.

I am a bit concerned that I am leaning towards too much information here.....
